In [2]:
import pandas as pd 
import numpy as np

In [3]:
data = pd.read_csv('train.csv')

In [4]:
data.head()

ID       y  X0 X1  X2 X3 X4 X5 X6 X8  ...  X375  X376  X377  X378  X379  \
0   0  130.81   k  v  at  a  d  u  j  o  ...     0     0     1     0     0   
1   6   88.53   k  t  av  e  d  y  l  o  ...     1     0     0     0     0   
2   7   76.26  az  w   n  c  d  x  j  x  ...     0     0     0     0     0   
3   9   80.62  az  t   n  f  d  x  l  e  ...     0     0     0     0     0   
4  13   78.02  az  v   n  f  d  h  d  n  ...     0     0     0     0     0   

   X380  X382  X383  X384  X385  
0     0     0     0     0     0  
1     0     0     0     0     0  
2     0     1     0     0     0  
3     0     0     0     0     0  
4     0     0     0     0     0  

[5 rows x 378 columns]

In [5]:
data.shape

(4209, 378)

In [6]:
data= pd.read_csv('train.csv', usecols=['X1', 'X2','X3','X4','X5','X6'])
data.head()

X1  X2 X3 X4 X5 X6
0  v  at  a  d  u  j
1  t  av  e  d  y  l
2  w   n  c  d  x  j
3  t   n  f  d  x  l
4  v   n  f  d  h  d

In [7]:
data['X6'].unique()

array(['j', 'l', 'd', 'h', 'i', 'a', 'g', 'c', 'k', 'e', 'f', 'b'],
      dtype=object)

In [8]:
# Let's have a look at how many labels each column (varaiable) has

for col in data.columns:
    print(col,':', len(data[col].unique()),'labels')

X1 : 27 labels
X2 : 44 labels
X3 : 7 labels
X4 : 4 labels
X5 : 29 labels
X6 : 12 labels


In [9]:
# let's examine how many columns we will p=obtion after one hot encoding these variables
pd.get_dummies(data,drop_first=True).shape

(4209, 117)

- We can see that from just 6 initial categorical variables, we end up with 117 new variables.
 - What can be done?
- In the winning solution of the KDD 2009 cup: "winnig the KDD Cup Orange Challenge with Enssemble Selection" the authr limit one hot encoding to the 10 most frequent labels of the variables. This means that they would make one binary variable for each of the 10 most frequent labels only. This is equivalent to grouping all the other labels under a new category, that in this case will be dropped. Thus, the 10 new dummy variables indcate if one of te\he 10 most frrequent labels is present (1) or not (0) for a particular observation.
    

In [10]:
# let's find the top 10 most frequent categories for the variabel X2
data.X2.value_counts().sort_values(ascending=False).head(20)

X2
as    1659
ae     496
ai     415
m      367
ak     265
r      153
n      137
s       94
f       87
e       81
aq      63
ay      54
a       47
t       29
k       25
i       25
b       21
ao      20
ag      19
z       19
Name: count, dtype: int64

In [11]:
data['X2'].value_counts().sort_values(ascending=False).head(20)

X2
as    1659
ae     496
ai     415
m      367
ak     265
r      153
n      137
s       94
f       87
e       81
aq      63
ay      54
a       47
t       29
k       25
i       25
b       21
ao      20
ag      19
z       19
Name: count, dtype: int64

In [12]:
# Let's take a list with the most frequent categories of the variable
top_10 = [x for x in data.X2.value_counts().sort_values(ascending =False).head(10).index] # index gives the categories names
top_10

['as', 'ae', 'ai', 'm', 'ak', 'r', 'n', 's', 'f', 'e']

In [13]:
# Now making we make the 10 binary variables

for label in top_10:
    data[label]=np.where(data['X2']== label,1,0)
    
data[['X2']+top_10].head(10)

X2  as  ae  ai  m  ak  r  n  s  f  e
0  at   0   0   0  0   0  0  0  0  0  0
1  av   0   0   0  0   0  0  0  0  0  0
2   n   0   0   0  0   0  0  1  0  0  0
3   n   0   0   0  0   0  0  1  0  0  0
4   n   0   0   0  0   0  0  1  0  0  0
5   e   0   0   0  0   0  0  0  0  0  1
6   e   0   0   0  0   0  0  0  0  0  1
7  as   1   0   0  0   0  0  0  0  0  0
8  as   1   0   0  0   0  0  0  0  0  0
9  aq   0   0   0  0   0  0  0  0  0  0

In [17]:
# Getting whole set of dummy variables, for all the categorical varaibles

def one_hot_top_x(data, variable, top_x_labels):
    #fucntion to create the dummy variables for the most frequent labels
    # We can vary the number of most frequent labels that we encode
    
    for label in top_x_labels:
        data[variable + '_'+ label] = np.where (data[variable]== label,1,0)
        
# read the data again
data= pd.read_csv('train.csv', usecols =['X1', 'X2','X3','X4','X5','X6'])

# encode X2 into the 10 most frequent categories
one_hot_top_x(data, 'X2', top_10)
data.head()

X1  X2 X3 X4 X5 X6  X2_as  X2_ae  X2_ai  X2_m  X2_ak  X2_r  X2_n  X2_s  \
0  v  at  a  d  u  j      0      0      0     0      0     0     0     0   
1  t  av  e  d  y  l      0      0      0     0      0     0     0     0   
2  w   n  c  d  x  j      0      0      0     0      0     0     1     0   
3  t   n  f  d  x  l      0      0      0     0      0     0     1     0   
4  v   n  f  d  h  d      0      0      0     0      0     0     1     0   

   X2_f  X2_e  
0     0     0  
1     0     0  
2     0     0  
3     0     0  
4     0     0

##### One Hot Encoding of top variables

###### Advantages
- Staightforward to implement
- Does not require hours of variable exploration
- Does not expand massively the feature space (number of columns in the dataset)

###### Disadvantages 
- Does not add any information that may make the variable more predictive
-Does not keep the information of the ignored labeld

###### Note:
- Because it is not unusual that categorical variables have a few dominating categories and the remaining labels add moslty noise, this i a quite simple and straightforward approach that may be useful on many occasions.

